pan3d/dataset_viewer.py

```diff
@@ -249,56 +360,76 @@ def set_render_options(
 
-        with self.state:
-            self.plotter.clear()
-            args = dict(
-                cmap=self.state.render_colormap,
-                clim=self.builder.data_range,
-                scalar_bar_args=dict(interactive=True),
-            )
-            if self.state.render_transparency:
-                args["opacity"] = self.state.render_transparency_function
-
-            try:
-                mesh = self.builder.mesh
-            except Exception as exception:
-                self.state.ui_error_message = str(exception)
-                self.state.ui_loading = False
-                return
-
-            if self.state.render_scalar_warp:
-                mesh = mesh.warp_by_scalar()
-            self.actor = self.plotter.add_mesh(
-                mesh,
-                **args,
-            )
-            if len(self.builder.data_array.shape) > 2:
-                self.plotter.view_isometric()
-            else:
-                self.plotter.view_xy()
-
-            if self.plot_view:
-                self.ctrl.push_camera()
-                self.ctrl.view_update()
+        if self.state.render_scalar_warp:
+            mesh = mesh.warp_by_scalar()
+        self.actor = self.plotter.add_mesh(
+            mesh,
+            **args,
+        )
+        if len(self.builder.data_array.shape) > 2:
+            self.plotter.view_isometric()
+        else:
+            self.plotter.view_xy()
 
-            self.state.ui_loading = False
+        if self.plot_view:
+            self.ctrl.push_camera()
+            self.ctrl.view_update()
 
     def apply_and_render(self, **kwargs) -> None:
         """Asynchronously reset and update cached mesh and render to viewer's plotter."""
 
-        asyncio.run_coroutine_threadsafe(self.plot_mesh(), self.current_event_loop)
+        self.run_as_async(self.plot_mesh)
+
+    def run_as_async(
+        self,
+        function,
+        loading_state="ui_loading",
+        error_state="ui_error_message",
+        unapplied_changes_state="ui_unapplied_changes",
+    ):
+        async def run():
+            with self.state:
+                if loading_state is not None:
+                    self.state[loading_state] = True
+                if error_state is not None:
+                    self.state[error_state] = None
+                if unapplied_changes_state is not None:
+                    self.state[unapplied_changes_state] = False
+
+            await asyncio.sleep(1)
+
+            with self.state:
+                try:
+                    function()
+                except Exception as e:
+                    if error_state is not None:
+                        self.state[error_state] = str(e)
+                    else:
+                        raise e
+                if loading_state is not None:
+                    self.state[loading_state] = False
+
+        if self.current_event_loop.is_running():
+            asyncio.run_coroutine_threadsafe(run(), self.current_event_loop)
```

Comment:
what is `run()`?

Reply:
`run` is a local function defined in `run_as_async`, just above this line. `run` acts as an aynchronous wrapper to an otherwise synchronous function (like `submit` or `plot_mesh`, which have had all their asynchronous parts moved to this wrapper). The `run` local function manages load state and error state changes that must be synchronized with the ui before starting a potentially long-running task. This was my way of consolidating all asynchronous logic to one reusable function instead of repeating myself in several places in the viewer class.